internal/controllers/synthesis/lifecycle.go

```diff
@@ -392,10 +393,46 @@ func shouldSwapStates(synth *apiv1.Synthesizer, comp *apiv1.Composition) bool {
 	// - synthesis is not already pending
 	// - all bound input resources exist and are in lockstep (or composition is being deleted)
 	syn := comp.Status.CurrentSynthesis
-	return (syn == nil ||
-		syn.ObservedCompositionGeneration != comp.Generation ||
-		(!inputRevisionsEqual(synth, comp.Status.InputRevisions, syn.InputRevisions) && syn.Synthesized != nil && !comp.ShouldIgnoreSideEffects())) &&
-		(comp.DeletionTimestamp != nil || (comp.InputsExist(synth) && !comp.InputsOutOfLockstep(synth)))
+
+	isSynNil := syn == nil
+	isCompGenDiff := syn != nil && syn.ObservedCompositionGeneration != comp.Generation
+	isInputRevisionsEqual := syn != nil && inputRevisionsEqual(synth, comp.Status.InputRevisions, syn.InputRevisions)
+	isSynthesizedNotNil := syn != nil && syn.Synthesized != nil
+	shouldIgnoreSideEffects := comp.ShouldIgnoreSideEffects()
+	isCompDeleted := comp.DeletionTimestamp != nil
+	isCompInputsExist := comp.InputsExist(synth)
+	isCompInputsOutOfLockstep := comp.InputsOutOfLockstep(synth)
+
+	reason := ""
+	if isSynNil {
```

Comment:
We don't need to accumulate all of the reasons, it's fine to just return the first one since they're already evaluated in a sensible order.

That can simplify the implementation a bit- you can perform each check and return on failure like:

```go
if !comp.InputsExist(synth) {
	return logger.WithValues(...), false
}
```

Reply:
The final result is calculated by AND first and second condition, and there are multiple OR condition inside first/second condition, so it can't tell the final result by checking only one condition.

From the example you mentioned, need to check "comp.DeletionTimestamp == nil" also to know the final result is false
```go
# second condition
(comp.DeletionTimestamp != nil || (comp.InputsExist(synth) && !comp.InputsOutOfLockstep(synth)))
```

I made the change to log the each result meet the condition by order instead of accumulate all the result meet the condition, but it's difficult to separate it by log only one condition, see the example log in next comment. Any better idea?

Reply:
Ahh I see the problem - we need to invert the logic

For example:

```go
if syn.ObservedCompositionGeneration != comp.Generation {
  return "CompositionModified", true
}

if !inputRevisionsEqual(synth, comp.Status.InputRevisions, syn.InputRevisions) {
  return "InputsChanged", true
}
```